src/shared/start-gateway.js

```diff
@@ -63,6 +63,36 @@ export async function startGateway<
         logger.info(`${name} running at http://${host}:${port}`);
     });
 
+    /**
+     * In this server is being run using a process manager, such as PM2, we
+     * may be asked to shutdown gracefully. We do this be listening for the
+     * SIGINT signal and then close the server. This prevents new connections
+     * from coming in and waits until the existing connections complete before
+     * the callback is fired. At which point we can safely shutdown the server.
+     * If we fail to respond then the process manager may try to forcefully
+     * shutdown the server after a timeout.
+     */
+    process.on("SIGINT", () => {
+        if (!gateway) {
+            return;
+        }
+
+        logger.info("SIGINT received, shutting down server.");
+
+        gateway.close((err) => {
+            if (err) {
+                logger.error(
+                    `Error shutting down server: ${
+                        (err && err.message) || "Unknown Error"
+                    }`,
+                );
+                process.exit(1);
+            } else {
+                process.exit(0);
+            }
+        });
```

Comment:
Should there be a try/catch around this too, just to make sure we catch all possible issues?